<a href="https://colab.research.google.com/github/kmblincow/CSC2300/blob/main/Week_4_Tutorial_Data_Wrangling_in_R.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# <font color='orange'>**Before you do anything else, make a copy of this notebook in your personal drive by going to "File > Save a copy in Drive" !!!!**
#**You should complete the tutorial using the copy you just made.**  </font>


---



---

#**Week 4 Tutorial - Data Wrangling in R**
In this tutorial, we will practice using the data wrangling tools that exist in the Tidyverse's dplyr package.


SOME IMPORTANT REMINDERS/NOTES:

-   Run code chunks by clicking the "play" sign on the left hand side, or pressing ctrl+Enter (PC) or cmd+Enter (Mac).  

-   You might receive some warning outputs in this tutorial related to NAs in the data set. Don't worry about that for right now.

- In this assignment, you will need to complete all of the "YOUR TURN" sections to receive full credit.
  


---



<div >
<img src="https://drive.google.com/uc?id=1f1ixDA3pZk5n_cSTu-RNdWf8-mOJMdpI" width=200/>
<img src="https://drive.google.com/uc?id=1hVSMXearF4oAkKWWxYNaJrcDhZj1nUPl" width=750/>
</div>

# **Data Wrangling using dplyr**

Note: This tutorial is largely created by Kevin Markham--I added some exercises for students to work through.

dplyr is a package from the Tidyverse that makes data wrangling much more intuitive and fun!

### Why use dplyr?

-   Great for data exploration and transformation
-   Intuitive to write and easy to read, especially when using the "chaining" or "piping" syntax (covered below)
-   Fast on data frames

### dplyr functionality

-   Five basic verbs: `filter`, `select`, `arrange`, `mutate`, `summarise` (plus `group_by`)

#**Our Data: MyAnimeList.net**

![](https://drive.google.com/uc?id=1064FLkNHPxqzj3-7aGAksLmBw7YLutjn)
    

This week's data comes from Tam Nguyen and MyAnimeList.net via Kaggle and TidyTuesday. According to Wikipedia -

> "MyAnimeList, often abbreviated as MAL, is an anime and manga social networking and social cataloging application website. The site provides its users with a list-like system to organize and score anime and manga. It facilitates finding users who share similar tastes and provides a large database on anime and manga. The site claims to have 4.4 million anime and 775,000 manga entries. In 2015, the site received 120 million visitors a month."
  
##Metadata

|variable       |class     |description |
|:--------------|:---------|:-----------|
|animeID        |double    | Anime ID (as in https://myanimelist.net/anime/animeID)          |
|name           |character |anime title - extracted from the site.           |
|title_english  |character | title in English (sometimes is different, sometimes is missing)          |
|type           |character | anime type (e.g. TV, Movie, OVA)          |
|source         |character | source of anime (i.e original, manga, game, music, visual novel etc.)         |
|producers      |character | producers          |
|genre          |character | genre         |
|studio         |character | studio           |
|episodes       |double    | number of episodes           |
|status         |character | Aired or not aired      |
|airing         |logical   | True/False is still airing          |
|start_date     |double    | Start date (ymd)        |
|end_date       |double    | End date (ymd)        |
|duration       |character | Per episode duration or entire duration, text string        |
|rating         |character | Age rating         |
|score          |double    | Score (higher = better)       |
|scored_by      |double    | Number of users that scored          |
|rank           |double    | Rank - weight according to MyAnimeList formula          |
|popularity     |double    |  based on how many members/users have the respective anime in their list          |
|members        |double    | number members that added this anime in their list         |
|favorites      |double    | number members that favorites these in their list          |
|premiered      |character | anime premiered on season/year          |


#**Getting Started**
It's good practice to always begin by loading the necessary packages for the code you will be writing. In our case, we will load the full suite of tidyverse packages using library(tidyverse).   
  
Make sure to run the code chunk below to accomplish that task.

In [ ]:
#load tidyverse packages (including dplyr)
library(tidyverse)

You should follow the steps from the Week 1 Tutorial:"Importing External Data" to load the anime.csv file from Brigthspace into this notebook. NOTE: you will have to reload the data if you leave and come back to this assignment, but the code will be saved.  
  
Once you have loaded the anime.csv file into your notebook, run the code chunk below to load it into your coding environment.

In [ ]:
#load our data into the coding environment
anime <- read.csv("/content/anime.csv")

<font color='deeppink'>**YOUR TURN:**  
Add some code to the chunk below that will show you a summary of your data.<font>

In [ ]:
#WRITE SOME CODE HERE TO EXPLORE YOUR NEWLY LOADED DATA


<font color='deeppink'>How many rows does the data have? How many columns?  



*DOUBLE CLICK TO INPUT YOUR ANSWER IN THIS TEXT CHUNK*  


#**filter(): Keep rows matching specific criteria**  
The filter() function in the dplyr package allows us to subset our data frame so that we only keep rows based on certain conditions.  
  
As a reminder from our early days doing the Lakers assignment, if we want to look for specific values we write the column name "==" the values we want. When filtering we use "==" not "=".
  
Let's begin by demonstrating how we can filter for rows associated with TV shows that were sourced from Manga.

In [ ]:
#use filter() to keep rows associated with TV shows that were sourced from Manga.
anime %>% filter(type == "TV" & source == "Manga")

Notice how the only values in the type column are "TV" and the source column are "Manga"? We kept only rows where those conditions were true.  
NOTE: we can use comma or ampersand (&) to represent the AND condition.  
  
We can use the "|" symbol for the OR condition.  
For example, if we want to filter for rows where genre was "Adventure" **or** "Drama" we can do so using the code below.

In [ ]:
#use filter() to keep rows where the genre is Adventure OR Drama
anime %>% filter(genre == "Adventure" | genre == "Drama")

We can also combine conditions in a single filter call. For example, if we wanted to keep rows from TV shows sourced from Manga that were in the Adventure or Drama genres we can run the code chunk below.

In [ ]:
#filter for Managa sourced Adventure or Drama TV shows
anime %>%
  filter(type == "TV" & source == "Manga" & (genre == "Adventure" | genre == "Drama"))

NOTE: We haven't saved any of these filtered subsets of the data, because we haven't assigned them a new object using the assignment arrow. If you wanted to save a filtered dataset to be used later, you would need to assign it to an object.

<font color='deeppink'>**YOUR TURN:**  
Add code to the chunk below that filters the anime data so you only keep rows that are TV shows or Movies that are ranked in the top 50 programs using the MyAnimeList formula. (HINT: you might want to use a less than (<) condition).


In [ ]:
#WRITE CODE HERE

#**select(): Pick columns by name**  
  
The select() function allows us to select specific columns. We can select columns by writing out their names directly, but there are also lots of helpful functions and shorthand operators that speed this process up.  
  
We can use the colon (:) to select multiple contiguous columns, as you can see in the code chunk below.

In [ ]:
#select all columns from title_english to genre
anime %>% select(title_english:genre)

We can also link column selections using commas. For example, we can select the name column and all columns between rank and favorites using the code below.

In [ ]:
#select name and all columns between rank and favorites
anime %>% select(name, rank:favorites)

There are also a number of helper functions that work with select that allow us to find columns that have names meeting certain conditions. These include: contains(), starts_with(), ends_with(), and matches(). For example, we can select all columns that contain the word "date" using the code chunk below.

In [ ]:
#select all columns with the word "date"
anime %>% select(contains("date"))

We can also remove columns using the "-" symbol. This is useful is you want to keep most of your columns and only drop one or two of them. For example, the "X" column is an artifact of loading the csv file. We can get rid of it using the code chunk below.

In [ ]:
#remove the X column
anime %>% select(-X)

You can combine any of the methods detailed above to pick and choose the columns you want to keep, just link them together using a comma.

<font color='deeppink'>**YOUR TURN:**  
Select all of the columns that have to do with the anime itself, excluding any columns that are related to MyAnimeList.net users and/or statistics.

In [ ]:
#WRITE CODE HERE

#**"Chaining" or "Piping"**  
  
Typically if you wanted to perform mutiple operations in one go in R you would need to nest functions within each other. This is confusing and difficult to read.  
  
The pipe operators (%>% or |>) allow you to write commands in a natural order that is more intuitive and easier to read. You can imagine the pipe operator to mean "and then".  
  
For example, say we wanted to select the name and rank columns, then filter so we only have titles that are ranked in the top 10, and then only show rows with distinct values for those column. We can accomplish that using the nesting syntax using the code chunk below.

In [ ]:
#accomplish multiple tasks at once using nesting
distinct(filter(select(anime, name, rank), rank < 11))

That's all well and good, but it's hard to force your brain to work from the inside of that code out following the nesting structure.  
A more intuitive way to accomplish the same task is to use pipe operators to sequentially accomplish the same tasks as shown in the code chunk below.

In [ ]:
#accomplish multiple tasks at once using piping
anime %>% #"take our anime data and then..."
  select(name, rank) %>% #"select the name and rank columns and then..."
  filter(rank < 11) %>% #"filter so we only have the top 10 titles and then..."
  distinct() #"only show rows with distinct combinations values from all columns"

You can hopefully see how piping increases readability significantly when there are many commands.

<font color='deeppink'>**YOUR TURN:**  
Using piping select the columns indicating the name of the anime, the type of anime, it's start date, and it's end date. Then filter so you only have TV shows. Then use the distinct() function to remove duplicate entries.

In [ ]:
#WRITE YOUR CODE HERE

#**arrange(): Reorder rows**  
The arrange() function allows us to order rows based on the value of a column or columns.  
For example, we might want to arrange the rows so our data is ordered based on the ranking of the anime. We can accomplish that using the code chunk below.

In [ ]:
#arrange data based on MyAnimeList.net ranking
anime %>% arrange(rank)

We can use the desc() function to arrange things in descending order. For example, we can use the code chunk below to arrange our data so that the anime that have been favorited the most are at the top.

In [ ]:
#arrange data in descending order of the favorites column
anime %>% arrange(desc(favorites))

<font color='deeppink'>**YOUR TURN:**  
Use piping in the code chunk below to select the name, rank, and popularity columns, then filter the data so that you only have the top 10 ranked anime. Then use the distinct() function to remove duplicate rows, and arrange the resulting data so that it's ordered with the most popular anime on top.

In [ ]:
#WRITE YOUR CODE HERE

#**mutate(): Add new variables**  
The mutate() function allows us to add new variables to a data frame. These variables can be standalone, or they can rely on existing variables.  
  
As an example what I mean by a "standalone" new variable, you can see in the code chunk below that I am creating a new column called "one" that is populated with the value 1.

In [ ]:
#add a column called "one" that is filled with 1s
anime %>% mutate(one = 1)

[1]     1 15468

But what is cooler, is that we can use values from other columns to perform calculations for our new column. For example, let's say we want to scale the popularity metric to be from 0 to 1, because it's hard to wrap our head around what 15468 popularity means (FYI that's the highest popularity score). We can do this by dividing all the popularity scores by the max popularity score, as you can see in the code chunk below.

In [ ]:
#create a new column that is the scaled popularity score
anime %>% mutate(popularity_scaled = popularity/max(popularity))

Note that all the code we have been running thus far is still just printing our manipulations and not saving them. If we want to store our new column, we need to use the assignment arrow, as seen in the code chunk below.

In [ ]:
#overwrite our original data with the data that includes the new popularity_scaled column
anime <- anime %>% mutate(popularity_scaled = popularity/max(popularity))

<font color='deeppink'>**YOUR TURN:**  
Scale the favorites column so that it ranges from 0 to 1, and store the resulting column by overwriting the anime object.

In [ ]:
#WRITE YOUR CODE HERE

#**group_by() & summarize(): Reduce variables to summary values**  
Many times we want to get summaries of our data. We can do that using summarize() (or summarise(), both work). We can use summarize() on it's own to get summary values for specific columns, as shown in the code chunk below where we calculate the mean number of episodes of anime.

In [ ]:
#calculate mean number of episodes
anime %>% summarize(mean_episodes = mean(episodes, na.rm = T))

NOTE: We need to add the "na.rm = T" argument to the mean function, because there are NAs in this column and R doesn't know how to handle that. We will learn more about NAs next week.  
  
Importantly, you see that summarize() completely changes the data frame. It collapses it into a summary of a single column. We can also link multiple summaries using commas in the summarize() function. We will demonstrate this in the code chunk below where we calculate the mean, min, and max number of episodes.

In [ ]:
#calculate mean, min, and max number of episodes
anime %>% summarize(mean_episodes = mean(episodes, na.rm = T),
                    min_episodes = min(episodes, na.rm = T),
                    max_episodes = max(episodes, na.rm = T))

You see that now our summary has three columns, one for each of the summary statistics we calculated (mean, min, and max).  
  
    
Often summarize() is used in conjunction with group_by(). group_by() creates an implicit grouping in your data frame based on a categorical variable. When summarize() comes after group_by(), the summary will be calculated at each level of the grouping variable.  
  
For example, in the code chunk below we will calculate the mean, min, and max number of episodes for each studio putting out anime.

In [ ]:
#calculate mean, min, and max episodes by genre
anime %>%
  group_by(studio) %>%
  summarize(mean_episodes = mean(episodes, na.rm = T),
            min_episodes = min(episodes, na.rm = T),
            max_episodes = max(episodes, na.rm = T))

Now you can see we have a row for each studio and columns showing the summary values for each studio.  
  
There are also helper functions for summarize(). These include n() which counts the number of rows in each group and n_distinct() which counts the number of unique items in a given vector.  
We might use these to count the number of anime made by each studio, and the number of distinct producers for each studio, as shown in the code chunk below.

In [ ]:
#print the number of anime made by each studio and number of distinct producers for each studio
anime %>%
  group_by(studio) %>%
  summarize(n_anime = n(),
            n_producers = n_distinct(producers))

<font color='deeppink'>**YOUR TURN:**  
Use the code chunk below to print data that shows the total number of anime in each genre, and the mean, min, and max score values by genre. Use piping to then arrange the data from highest to lowest max score value.

In [ ]:
#WRITE YOUR CODE HERE


#**Additional Resources**  
-   [Official dplyr reference manual and vignettes on CRAN](http://cran.r-project.org/web/packages/dplyr/index.html): vignettes are well-written and cover many aspects of dplyr
-   [dplyr GitHub repo](https://github.com/hadley/dplyr) and [list of releases](https://github.com/hadley/dplyr/releases)  
  
*This tutorial was created by Kayla Blincow for UVI's Data Science Program. If you have questions about it's implmementation please contact Dr. Blincow at kayla.israel@uvi.edu*.